File: python/matrix_operation.py
#!/home/ahmad/miniconda3/bin/python

import time

N = 1000


def create_matrix(n):
    return [[(i * n + j) % 100 for j in range(n)] for i in range(n)]

def matrix_mul(a, b, n):
    c = [[0.0 for _ in range(n)] for _ in range(n)]
    for i in range(n):
        for j in range(n):
            sum_ = 0.0
            for k in range(n):
                sum_ += a[i][k] * b[k][j]
            c[i][j] = sum_
    return c


def checksum(matrix):
    return sum(sum(row) for row in matrix)


def main():
    a = create_matrix(N)
    b = create_matrix(N)

    start = time.perf_counter()
    c = matrix_mul(a, b, N)
    end = time.perf_counter()

    print(f"Matrix size: {N} x {N}")
    print(f"Time taken: {end - start:.6f} seconds")
    print(f"Result checksum: {checksum(c):.2f}")

if __name__ == "__main__":
    main()

File: fortran/matrix_operation.f90
program matrix_mul
  implicit none
  integer, parameter :: dp = kind(1.0d0)
  integer, parameter :: n = 1000  ! Use 1000+ for benchmarks

  real(dp), allocatable :: a(:,:), b(:,:), c(:,:)
  integer :: i, j, k
  real(dp) :: start, finish, checksum

  allocate(a(n,n), b(n,n), c(n,n))

  ! Fill matrices with known values
  do i = 1, n
     do j = 1, n
        a(i,j) = mod((i - 1) * n + (j - 1), 100)
        b(i,j) = mod((i - 1) * n + (j - 1), 100)
     end do
  end do

  ! Print inputs (optional for small n)
  ! if (n <= 5) then
  !    print *, "Matrix A:"
  !    call print_matrix(a, n)
  !    print *, "Matrix B:"
  !    call print_matrix(b, n)
  ! end if

  ! Time only the matrix multiplication
  call cpu_time(start)
  do i = 1, n
     do j = 1, n
        c(i,j) = 0.0_dp
        do k = 1, n
           c(i,j) = c(i,j) + a(i,k) * b(k,j)
        end do
     end do
  end do
  call cpu_time(finish)

  ! Output
  print *, "Matrix size:", n, "x", n
  print *, "Time taken:", finish - start, "seconds"

  ! if (n <= 5) then
  !    print *, "Result matrix C:"
  !    call print_matrix(c, n)
  ! end if

  checksum = sum(c)
  print *, "Result checksum:", checksum

contains

  subroutine print_matrix(mat, n)
    real(dp), intent(in) :: mat(n,n)
    integer, intent(in) :: n
    integer :: i, j
    do i = 1, n
       write(*,'(100(f8.2,1x))') (mat(i,j), j=1,n)
    end do
  end subroutine print_matrix

end program matrix_mul

